### ISSUE TITLE
Predict during training

### ISSUE THREAD

**Issue by mikerossgithub** [allenai/allennlp#3218]
**Is your feature request related to a problem? Please describe.**
I would like to log model predictions for every epoch, even though I am only keeping checkpoints of the latest 3 epochs. This enables post-analysis on the predictions without the disk overhead of logging model weights every epoch. One use case is to generate training curves for new metrics on old (expensive) training runs.

**Describe the solution you'd like**
Either a Predict callback or a "predict" field added to the Validate callback

I looked into subclassing or modifying the Validate callback, but since I need the `output_dict`, I would also need to alter `CallbackTrainer.batch_loss()` to return the `output_dict` instead of just returning loss. For now, I have written a hack that duplicates CallbackTrainer.batch_loss() functionality inside the callback.

**Comment by matt-gardner** [comment]
This is an argument in favor of your original design @joelgrus, right?  Where basically everything was in callbacks?

**Comment by joelgrus** [comment]
that design was too complex, so I'm not sure this is an argument for it.

here's I think a possible solution: modify `CallbackTrainer.batch_loss` so that it saves `output_dict` as part of the trainer state, then any callback can use it for whatever it wants:

https://github.com/allenai/allennlp/blob/master/allennlp/training/callback_trainer.py#L152

would that solve your problem?

**Comment by mikerossgithub** [comment]
So CallbackTrainer would have a `most_recent_batch_output` variable? That would solve my particular problem! But two other changes would be nice. Neither of these are necessary, so if you don't think they are worth it, I would be happy with the solution you proposed!

1.
https://github.com/allenai/allennlp/blob/5e2206df306d3e11640032b523dc1a6f24a1c95b/allennlp/training/callbacks/validate.py#L61
This could be refactored to separate the somewhat non-trivial code for validation data traversal from the metric accumulation. Then the PredictCallback can just subclass and override the metric accumulation with prediction accumulation. As it stands, I need to copy this entire method into my PredictCallback and monitor it for future AllenNLP changes. 

2. 
Instead of `most_recent_batch_output`, one could just refactor batch_loss into `calc_loss(batch_forward(batch), for_training)` -- the PredictCallback can then get the output_dict with `batch_forward` without adding a transient state variable to the CallbackTrainer.

**Comment by sai-prasanna** [comment]
@matt-gardner  @joelgrus I think it might be worth looking into [PyTorch-lightning](https://github.com/williamFalcon/pytorch-lightning) Trainer. They support DDP, distributed training, gradient accumulation, callbacks, fp16 (with apex), checkpointing + restarts and unifrom logging across processes and experiment management system (for now test_tube).

**Comment by schmmd** [comment]
Related to https://github.com/allenai/allennlp/issues/3519
